fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1470,6 +1470,86 @@ def _free_full_params(self, params: Optional[List[Parameter]] = None) -> None:
+            return ".".join(
+                [split for split in path.split(".") if split not in {"_fsdp_wrapped_module", "_fpw_module"}]
+            )
+
+        fsdp_paths = []
+        param_names = []
+        num_padded = []
+        numels = []
+        unflat_shapes = []
+        no_broadcast_optim_state = []
+        for path, m in self.named_modules():
+            if isinstance(m, FullyShardedDataParallel):
+                assert hasattr(m, "_param_numels"), "This method only supports flatten_parameters=True"
+                fsdp_paths.append(clean_path(path))
+                fsdp_instance_param_names = []
+                for param_path, param_name in m._param_full_infos:
+                    full_param_path = param_path + "." + param_name if param_path else param_name
+                    fsdp_instance_param_names.append(clean_path(full_param_path))
+                param_names.append(fsdp_instance_param_names)
+                num_padded.append(m.numel_padded_per_param)
+                numels.append(m._param_numels)
+                unflat_shapes.append(m._param_shapes)
+                no_broadcast_optim_state.append(m.no_broadcast_optim_state)
+
+        return dict(
+            fsdp_paths=fsdp_paths,
+            param_names=param_names,
+            num_padded=num_padded,
+            numels=numels,
+            unflat_shapes=unflat_shapes,
+            no_broadcast_optim_state=no_broadcast_optim_state,
+        )
+
+    @staticmethod
+    def consolidate_shard_weights(
+        shard_weights: List[Dict[str, torch.Tensor]],
+        shard_metadata: List[Dict[str, Any]]
+    ):
+        """
+        Given a list of weights and meta data associated to N shards, reconstruct
+        the weights of an equivalent unsharded model
+
+        This method is very useful to re-assemble checkpoints of shards without
+        having to instantiate FSDP wrappers with the world size originally used
+        to save the shards
+        """
+        assert len(shard_weights) > 0, "Invalid number of shards: 0"
+        assert len(shard_weights) == len(shard_metadata), "Require meta data for each shard"
+
+        consolidated_weights = {}
+        original_world_size = len(shard_weights)
+
+        num_fsdp_wraps = len(shard_metadata[0]["fsdp_paths"])
+        for i in range(num_fsdp_wraps):
+            fsdp_path = shard_metadata[0]["fsdp_paths"][i]
+            fsdp_param_name = ".".join([fsdp_path, "flat_param"]) if fsdp_path else "flat_param"
+            shards = []
+            for rank in range(original_world_size):
+                shard = shard_weights[rank][fsdp_param_name]
+                pad = shard_metadata[rank]["num_padded"][i][0]
+                if pad > 0:
+                    shard = shard[:-pad]
+                shards.append(shard)
+            full_param = torch.cat(shards, dim=0)
+
+            param_names = shard_metadata[0]["param_names"][i]
+            param_numels = shard_metadata[0]["numels"][i]
```

Comment:
did we check sum(param_numels) equals to the full_param numel? If not, I think it is a good idea to check and raise ValueError.

Reply:
In general, since the data is potentially from the disk, we better check its integrity as much as we can. Feel free to add more checks. Better die early then later in some downstream place where is potentially harder to debug.